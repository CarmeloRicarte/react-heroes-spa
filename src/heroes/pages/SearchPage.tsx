import { HeroCard } from "../components";
import { useForm } from "../../hooks";
import { FormEvent } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";
import { getHeroesByName } from "../helpers";
interface FormData {
  searchText: string;
}
export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search); // divide query params
  const heroes = getHeroesByName(q + "");
  const showSearch = q?.length === 0;
  const showError = q && q?.length > 0 && heroes.length === 0;
  const { searchText, onInputChange } = useForm<FormData>({
    searchText: q + "",
  });

  const onSearchSubmit = (
    event:
      | FormEvent<HTMLFormElement>
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    event.preventDefault();
    navigate(`${location.pathname}?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              data-testid="search-input"
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
          </form>

          <button
            data-testid="search-button"
            type="submit"
            onClick={onSearchSubmit}
            className="btn btn-outline-primary mt-3"
          >
            Search
          </button>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          <div
            className="alert alert-primary animate__animated animate__fadeIn"
            style={{ display: showSearch ? "" : "none" }}
          >
            Search a hero
          </div>
          <div
            className="alert alert-danger animate__animated animate__fadeIn"
            style={{ display: showError ? "" : "none" }}
          >
            No hero with <b>{q}</b>
          </div>

          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
