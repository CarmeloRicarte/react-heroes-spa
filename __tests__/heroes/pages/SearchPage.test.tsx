import { describe, test, expect, afterEach, vi } from "vitest";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import React from "react";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";
const mockedUseNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual<typeof import("react-router-dom")>(
    "react-router-dom"
  );
  return {
    ...mod,
    useNavigate: () => mockedUseNavigate,
  };
});
describe("Tests of <SearchPage />", () => {
  afterEach(() => {
    cleanup();
    vi.clearAllMocks();
  });

  test("should show correctly with default value", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test("should show Batman and the input with the same value as queryString", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    );
    const input = screen.getByTestId("search-input") as HTMLInputElement;
    const img = screen.getByRole("img") as HTMLImageElement;
    expect(input.value).toBe("batman");
    expect(img.src).toContain("/assets/heroes/dc-batman.jpg");
    expect(screen.getByTestId("searchAHeroDiv").style.display).toBe("none");
  });

  test("should show an error if the hero is not found(batman123)", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <SearchPage />
      </MemoryRouter>
    );
    expect(screen.getByTestId("noHeroAlert").style.display).not.toBe("none");
  });

  test("should call navigate to new screen", () => {
    // ingress value
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>
    );
    const input = screen.getByTestId("search-input") as HTMLInputElement;
    fireEvent.input(input, { target: { value: "hulk" } });
    // fire submit event with button that have a submit event
    fireEvent.click(screen.getByTestId("search-button"));
    // get input value to call navigate with this argument
    expect(mockedUseNavigate).toHaveBeenCalledWith(`/search?q=${input.value}`);
  });
});
