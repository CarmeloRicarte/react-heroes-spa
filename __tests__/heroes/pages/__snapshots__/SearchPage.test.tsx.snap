// Vitest Snapshot v1

exports[`Tests of <SearchPage /> > should show correctly with default value 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          data-testid="search-input"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
      </form>
      <button
        class="btn btn-outline-primary mt-3"
        data-testid="search-button"
        type="submit"
      >
        Search
      </button>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
        data-testid="searchAHeroDiv"
      >
        Search a hero
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        data-testid="noHeroAlert"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
